client/utils/src/pubsub.rs

```diff
@@ -0,0 +1,246 @@
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+//! Provides means to implement a typical Pub/Sub mechanism.
+//!
+//! This module provides a type [`Hub`] which can be used both to subscribe,
+//! and to send the broadcast messages.
+//!
+//! The [`Hub`] type is parametrized by two other types:
+//! - `Message` — the type of a message that shall be delivered to the subscribers;
+//! - `Registry` — implementation of the subscription/dispatch logic.
+//!
+//! A Registry is implemented by defining the following traits:
+//! - [`Subscribe<K>`];
+//! - [`Dispatch<M>`];
+//! - [`Unsubscribe`].
+//!
+//! As a result of subscription `Hub::subscribe` method returns an instance of
+//! [`Receiver<Message,Registry>`]. That can be used as a [`Stream`] to receive the messages.
+//! Upon drop the [`Receiver<Message, Registry>`] shall unregister itself from the `Hub`.
+
+use std::{
+	collections::HashMap,
+	ops::DerefMut,
+	pin::Pin,
+	sync::{Arc, Weak},
+	task::{Context, Poll},
+};
+
+use futures::stream::{FusedStream, Stream};
+use parking_lot::{Mutex, MutexGuard};
+
+use crate::{
+	id_sequence::SeqID,
+	mpsc::{TracingUnboundedReceiver, TracingUnboundedSender},
+};
+
+/// Unsubscribe: unregisters a previously created subscription.
+pub trait Unsubscribe {
+	/// Remove all registrations of the subscriber with ID `subs_id`.
+	fn unsubscribe(&mut self, subs_id: SeqID);
+}
+
+/// Subscribe using a key of type `K`
+pub trait Subscribe<K> {
+	/// Register subscriber with the ID `subs_id` as having interest to the key `K`.
+	fn subscribe(&mut self, subs_key: K, subs_id: SeqID);
+}
+
+/// Dispatch a message of type `M`.
+pub trait Dispatch<M> {
+	/// The type of the that shall be sent through the channel as a result of such dispatch.
+	type Item;
+	/// The type returned by the `dispatch`-method.
+	type Ret;
+
+	/// Dispatch the message of type `M`.
+	///
+	/// The implementation is given an instance of `M` and is supposed to invoke `dispatch` for
+	/// each matching subscriber, with an argument of type `Self::Item` matching that subscriber.
+	///
+	/// Note that this does not have to be of the same type with the item that will be sent through
+	/// to the subscribers. The subscribers will receive a message of type `Self::Item`.
+	fn dispatch<F>(&mut self, message: M, dispatch: F) -> Self::Ret
+	where
+		F: FnMut(&SeqID, Self::Item);
+}
+
+/// A subscription hub.
+///
+/// Does the subscription and dispatch.
+/// The exact subscription and routing behaviour is to be implemented by the Registry (of type `R`).
+/// The Hub under the hood uses the channel defined in `crate::mpsc` module.
+#[derive(Debug)]
+pub struct Hub<M, R> {
+	tracing_key: &'static str,
+	shared: Arc<Mutex<Shared<M, R>>>,
```

Comment:
If either `Hub::subscribe` or `Hub::send` gets called or another `Receiver` from the same `Hub` is dropped on the same thread from within `Subscribe::subscribe`, `Unsubscribe::unsubscribe` or `Dispatch::dispatch` then this mutex will deadlock.

1. We should detect when that happens and `panic!` instead of deadlocking. (e.g. by keeping a `ThreadLocal` from the [`thread_local`](https://crates.io/crates/thread_local) crate tracking whether the `Mutex` was already locked on the current thread and panicking if it was)
2. It should be documented that calling those methods in those situations is disallowed.
3. A test for the panicking behavior would be nice.

Reply:
As I see it, such situation may occur if a Mutex's locked state spans across several `Future::poll` calls.

Every call that touches the Hub's mutex — is a sync-call, i.e. such call shall not return without releasing the mutex.

Did I understand correctly what your concerns were?

Reply:
I meant in general, not only when a `Mutex` is locked across multiple `await`s.

If you lock a Mutex then you can't lock it again from the same thread, or it will permanently deadlock. (And deadlocks are not fun to debug.) Currently it's possible to trigger such a deadlock, which isn't great. Instead of allowing for a deadlock we should just detect such cases and panic.

Reply:
Isn't that a bit overly protective?

That particular mutex is not a part of a public interface<sup>*</sup>, hence neither the outside-user of a `Hub`, nor the implementor of this module's traits can violate the assumptions of how this mutex is used.

If that protective attitude to the private implementation details is a common practice here — I am prepared to do as you tell, of course. 

<sup>*</sup> given that I will mark `fn lock_registry(...)` as `#[cfg(test)]`.

Reply:
> hence neither the outside-user of a `Hub`, nor the implementor of this module's traits can violate the assumptions of how this mutex is used.

That's incorrect. (: `Mutex` assumes that `lock()` will be called only once from a single thread, and that assumption is indirectly also part of our interface here. It can be deadlocked from outside just by calling the public interface in a certain way (even if you'd delete the currently public `lock_registry` that will be possible), so it's not just a private implementation detail, it's a potential footgun.

Reply:
We can also just use: https://docs.rs/parking_lot/latest/parking_lot/type.ReentrantMutex.html

Reply:
> We can also just use: https://docs.rs/parking_lot/latest/parking_lot/type.ReentrantMutex.html

Ah, indeed! That would probably be simpler! :+1:

Reply:
@koute , @bkchr , would you be able to draw an unambiguous "line" dividing the cases when it is safe to use just a `Mutex` in this project, and when the gymnastics with a `ReentrantMutex` + `RefCell` would be due?

I am not being stubborn, I genuinely do not see a certain way to just call the public API, so that the `Mutex` remains sideways afterwards.

Reply:
> would you be able to draw an unambiguous "line" dividing the cases when it is safe to use just a `Mutex` in this project, and when the gymnastics with a `ReentrantMutex` + `RefCell` would be due?

In no case it should be possible to inadvertently deadlock a `Mutex` when you're not explicitly handling a `Mutex`.

In this case, again, you can trigger a deadlock of this internal `Mutex` *from the outside*. You don't have to directly access it. You don't have to even know that it's in there.

> I genuinely do not see a certain way to just call the public API, so that the `Mutex` remains sideways afterwards.

Let me quote myself:

> If either `Hub::subscribe` or `Hub::send` gets called or another `Receiver` from the same `Hub` is dropped on the same thread from within `Subscribe::subscribe`, `Unsubscribe::unsubscribe` or `Dispatch::dispatch` then this mutex will deadlock.

So, for example:

1. A `Receiver` is dropped.
2. `Drop::drop` is called for the `Receiver`.
3. The `Mutex` is locked from inside of `Receiver`'s `drop`.
4. `Shared::unsubscribe` is called while the `Mutex` is being locked.
3. `<R as Unsubscribe>::unsubscribe` is called.
4. The `unsubscribe` impl drops another `Receiver`, either explicitly or entirely by accident due to some weird machination.
5. The `Mutex` is locked again; the previous lock is still in effect.
6. We deadlock.

Reply:
I am having hard time imagining how step-6 could possibly happen.
It would be required for a Registry to possess an instance of a `Receiver` (of the same hub).
Wouldn't that be exotic to such extent that it would not pass a review such as this one?

Reply:
It doesn't have to be directly, it might be indirectly due to some weird indirect machination, e.g. the registry could posses an `Arc` to something, and that object could contain another object, and that another object could contain a `Receiver` (e.g. maybe it would be a list of other receivers or something), and suddenly that `Arc` is the last one out there so the whole thing gets dropped.

It doesn't really matter; what's important is that it's possible to trigger this by mistake. You could use the same argument as for why we might as well use C++ and not Rust: "but writing out of bounds of this array would not pass a review such as this one, so why have bounds checking?". It doesn't matter; the code has to be correct in every circumstance, because sooner or later somebody will screw up. (Same reason why scoped threads were removed from Rust's standard library. You had to really work for it to trigger the unsafety so it was unlikely for anyone to trigger it, but it was unsound so it was removed.)

Reply:
Okay. I will add the reentrant+refcell thing here,
although I am still convinced that in order to become a footgun that would require quite a lot of dubious practices to be used.

Thanks.